Split IO #2

francisco-dlp opened the pull request:
### Description of the change

Split HyperSpy's IO plugins into a separate package RosettaSciIO. See #1978 for the related discussion.

This supersedes #2174

### Progress of the PR
- [ ] Change implemented (can be split into several points)
    - [x] Move all plugins to a separate folder, with its own setup.py, that must be installed separately.
    - [x] Move each plugin to a folder of its own
    - [x] Move all IO tests to rsciio
    - [x] Enable lazy loading of the plugins by moving the specifications to yamls files.
    - [x] Make rosettasciio fully independent of HyperSpy
        - [x] Don't use ``DictionaryTreeBrowser``. DTB was replaced by python-box
        - [x] Move ``misc.io`` code to ``rsciio``
        - [x] Move exceptions to ``rsciio``. 
        - [x] Move remaining imports to ``rsciio``Most of it is done, except for:
            - [x] ``rgb_tools``
            - [x]  ``emd`` relies on hyperspy for reading
        - [x] Make `_hierarchical` fully independent of HyperSpy
        - [x] Pass dictionary instead of signal to ``file_writer``s
            - [x] msa
            - [x] tvips
            - [x] semper
            - [x] blockfile
            - [x] emd
            - [x] panta rhei
            - [x] nexus
            - [x] image (https://github.com/francisco-dlp/hyperspy/pull/51)
            - [x] mrcz (https://github.com/francisco-dlp/hyperspy/pull/51)
            - [x] ripple (https://github.com/francisco-dlp/hyperspy/pull/51)
            - [x] nexus (https://github.com/francisco-dlp/hyperspy/pull/51)
            - [x] tiff (https://github.com/francisco-dlp/hyperspy/pull/51)
            - [x] usid (https://github.com/francisco-dlp/hyperspy/pull/51)
            - [x] zspy
            - [x] hspy
    - [x] Move User Guide IO chapter to rsciio
    - [x] Remove dependency on traits (basically for t.Undefined)
    - [ ]  Make tests independent of hyperspy (not strictly required for release)
- [ ] update docstring (if appropriate),
- [x] update user guide (if appropriate),
- [x] add an changelog entry in the `upcoming_changes` folder (see [`upcoming_changes/README.rst`](https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/upcoming_changes/README.rst)) -- see #2979, 
- [ ] Check formatting changelog entry in the `readthedocs` doc build of this PR (link in github checks)
- [ ] add tests,
- [x] ready for review.

### How to test it

To test it, simply install `rosettasciio` (currently at hyperspy's root folder, it'll move to its own repository once there are no more code exchanges between hyperspy and RosettaSciIO) in addition to hyperspy. Everything should work as usual.

Additionally, it is possible to import the readers directly from RosettaSciIO as follows:

```python
from rsciio.msa import api
api.file_reader("your_msa_file.msa")
```

TomSlater committed:
Updated the file_reader docstring to include reading of 'de5' files.

TomSlater committed:
Add de5 to io.rst

Commit:
Add io_plugin for CEOS Panta Rhei *.prz file format

prz files are just npz files, containing an array, and a dictionary with
the relevant metadata.

Commit:
Add relevant docstrings to public functions

Commit:
Add changelog entry

Commit:
Add test for PRZ file reader

Minor: enable compression in PRZ file writer

Commit:
Add user documentation of PRZ file support.

Commit:
Update test for PRZ file reader.

Due to changes in the hyperspy metadata structure the old test was failing.

francisco-dlp committed:
Move io_plugins to new package folder rosettascio

francisco-dlp committed:
Add setup, license and README

francisco-dlp committed:
Update imports

francisco-dlp committed:
Move io tests to rsciio

francisco-dlp committed:
Fix import

francisco-dlp committed:
Add untracked zip folder

francisco-dlp committed:
Move each plugin to each own folder and rename all to api.py

francisco-dlp committed:
Add __init__.py

francisco-dlp committed:
Move io specifications to yaml file

francisco-dlp committed:
Use yaml files to create IO_PLUGINS list

francisco-dlp committed:
Use yaml files to create IO_PLUGINS list

francisco-dlp committed:
Remove IO from setup.py

francisco-dlp committed:
Some fixes

francisco-dlp committed:
test_io passing

francisco-dlp committed:
Fix some issues

francisco-dlp committed:
Raise an error when multiple readers can read a file

francisco-dlp committed:
Add file_format keyword to save and raise error when multiple formats possible

francisco-dlp committed:
Fix test_hdf5

francisco-dlp committed:
Fix tiff tests

francisco-dlp committed:
Fix USID

francisco-dlp committed:
Fix ZSPY

francisco-dlp committed:
Fix saving when multiple file_formats possible

francisco-dlp committed:
Fix cython bruker

francisco-dlp committed:
Move hyperspy/misc/io to rsciio/utils

francisco-dlp committed:
Replace DictionaryTreeBrowser with python-box in DM reader

francisco-dlp committed:
Replace DictionaryTreeBrowser with python-box in MSA reader

francisco-dlp committed:
Replace DictionaryTreeBrowser with python-box in FEI reader

francisco-dlp committed:
Replace DictionaryTreeBrowser with python-box in EMPAD reader

francisco-dlp committed:
Replace DictionaryTreeBrowser with python-box in ripple reader

francisco-dlp committed:
Move IO exceptions to rsciio

francisco-dlp committed:
Move IO exceptions to rsciio

francisco-dlp committed:
Remove hyperspy import from mrcz

francisco-dlp committed:
Remove hyperspy dependency from semper reader

francisco-dlp committed:
Merge branch 'SPLIT_IO-2' of https://github.com/francisco-dlp/hyperspy into SPLIT_IO-2

francisco-dlp commented:
This is still WIP, but I am submitting it now for feedback. AFAIK, everything works as it should, but RosettaSciIO is not yet fully independent from HyperSpy.

I suggest the following release strategy:
* Release it with all the current plugins once it is fully independent from HyperSpy (with the exception of the test, that could continue depending on HyperSpy)
* Once we split the EM code from HyperSpy, we can split the EM readers from RosettaSciIO, leaving only the generic `zspy` and `hspy` formats in.

To avoid difficult merges, I suggest to send all changes to the IO code (including new reader) to this branch. PRs to help with finishing this PR are more than welcome, but think on discussing here what you are planning to work on to avoid duplicating efforts.

jlaehne commented on rosettascio/rsciio/blockfile/specifications.yaml:
```diff
@@ -0,0 +1,7 @@
+file_extensions: [blo, BLO]
+default_extension: 0
+writes: [[2, 2], [2, 1], [2, 0]]
+non_uniform_axis: False
```
Complains about missing newlines for some of the `specifications.yaml` files

jlaehne commented:
Thanks for moving forward on this @francisco-dlp

francisco-dlp committed:
Move sarray2dict and dict2sarray to rsciio

francisco-dlp committed:
Move some dateutil functions to rsciio

francisco-dlp committed:
Fix nexus hyperspy import

francisco-dlp committed:
Fix bruker unbcf_fast import

francisco-dlp committed:
Add dummy_context_manager to rsciio

francisco-dlp committed:
Add missing files

francisco-dlp committed:
Remove hspy deprecation_warning from bruker

francisco-dlp committed:
Remove hspy optimize_arg from DM

francisco-dlp committed:
Add elements to rsciio

CSSFrancis commented:
@francisco-dlp Thank you so much for doing this!  I think this is going to be very helpful moving forward. I can look into making `_hierarchical` independent of hyperspy if you need help with that.  

All that seems to need to be done there is to move some of the type checking to the zspy and hspy files.  We could use the dictionary key in the `dict2group` function rather than checking the type directly but I fear that might break the reading of older hyperspy file formats.

francisco-dlp commented:
Thanks for offering help with this @CSSFrancis.

We certainly cannot break compatibility with older hyperspy file formats. In addition to what you mention, we must pass a dictionary rather than a hyperspy `Signal` to ``file_writer`` to make it fully hyperspy-free. This is true for all the write-capable plugins.

francisco-dlp commented:
@CSSFrancis, as I don't fully understand what you are proposing, even if it is obvious, just in case I mention that  the `zspy` and `hspy` readers must be fully independent of hyperspy too.

ericpre committed:
Fix CI

CSSFrancis commented:
> @CSSFrancis, as I don't fully understand what you are proposing, even if it is obvious, just in case I mention that the `zspy` and `hspy` readers must be fully independent of hyperspy too.

@francisco-dlp Thanks for the clarification,  I think I misread what you were describing above about leaving only the `.zspy` and `hspy` formats.  I'll play around with it and see what it takes to pass everything as a dictionary.

ericpre commented:
This is great!

> * [ ]  Create per plugin mapping yaml that maps the original metadata to hyperspy's metadata

Is it something that could be done later?

I can volunteer for doing the docs and sorting out what is left to do for the import. :)

ericpre committed:
Fix and tidy up setup.py

jlaehne commented:
> > [ ] Create per plugin mapping yaml that maps the original metadata to hyperspy's metadata
> 
> Is it something that could be done later?

I agree that it should not be a requirement for the split.

That being said, as we are currently developing several new readers, it would of course be nice to have that functionality available quite soon so that we don't have to adapt that part again later.

ericpre committed:
Fix import

ericpre committed:
Fix test

ericpre committed:
Fix test with reader inference for hdf5 file

francisco-dlp commented:
> @francisco-dlp Thanks for the clarification, I think I misread what you were describing above about leaving only the .zspy and hspy formats

@CSSFrancis, sorry that I wasn't clear enough about that. The point is that it should be possible to read/write all the supported formats (including `hspy` and `zspy`) without installing HyperSpy.

>> * [ ]  Create per plugin mapping yaml that maps the original metadata to hyperspy's metadata
>
>Is it something that could be done later?

Yes, I agree that we can leave that for later so that we can release HyperSpy v1.8 and RosettaSciIO asap. I have removed it from the progress tracker above.

jlaehne commented:
> Yes, I agree that we can leave that for later so that we can release HyperSpy v1.8 and RosettaSciIO asap. I have removed it from the progress tracker above.

Though we should then remember to add it to a tracking issue of next steps once the RosettaScIO repo has been created.

ericpre commented:
Made a PR to this branch to fix installation and the test suite: https://github.com/francisco-dlp/hyperspy/pull/48

francisco-dlp committed:
Merge pull request #48 from ericpre/SPLIT_IO-2_fix_test_suite

[SPLIT_IO] Fix installation and test suite

francisco-dlp commented:
Thanks @ericpre. I've just merged it.

ericpre commented:
I am wondering if we should try to do as little as required for this step and defer as much as possible in the separate repository, for example, the two following points could be done separately:

> * [ ]  Make `_hierarchical` fully independent of HyperSpy
> * [ ]  Pass dictionary instead of signal for `file_writer`s

by following these steps:

1. make a first rosettascio release without hyperspy release
2. update dependencies requirement in hyperspy development branches
   - Use it for some time to gain confidence that nothing is broken
   - address outsanding task in rosettascio
4. make another rosettascio release and hyperspy 1.8, once we are happy with everything.

At the same time, we could include #2803 and #2896, to avoid resubmitting these PRs in the new repository - these PRs are in good shape but are missing test files.

francisco-dlp commented:
I think that it is good to finish both of those tasks before splitting the repository. This is just for the sake of convenience, since they involve changing both, HyperSpy and RosettaSciIO's code. Also, we must keep the RosettaSciIO folder inside the hyperspy repository until we are sure that we don't have to move more files around. Otherwise we'll lose the git history of the file.

Your release plan sounds pretty sensible. Probably we'll need to keep RosettaSciIO in beta status (i.e. without API stability) for a while.

Regarding #2803 and #2896, I think that it is important that the authors add tests before merging them.

francisco-dlp committed:
Import ureg from rsciio

francisco-dlp committed:
Merge branch 'SPLIT_IO-2' of https://github.com/francisco-dlp/hyperspy into SPLIT_IO-2

francisco-dlp committed:
Add package version

francisco-dlp committed:
Replace hyperspy version with rsciio version

jlaehne commented:
Codecov unveiled a number of missing tests for possible errors raised by `io.py`: https://github.com/hyperspy/hyperspy/pull/2972/files/89ee1ea71cfceacbb1792928215637c6ede26eb0..2d93734c617bd7e9e92e1ff2c0b27aea027663d0

They should be low hanging fruits and it might make sense to get them covered along the way.

francisco-dlp committed:
Remove strong dependency from hyperspy.mist.rgb_tools

ericpre commented on rosettascio/rsciio/image/api.py:
```diff
@@ -80,9 +79,13 @@ def file_writer(filename, signal, scalebar=False, scalebar_kwds=None,
-        data = rgb_tools.rgbx2regular_array(data)
+    try:
+        # HyperSpy uses struct arrays to store RGBA data
+        from hyperspy.misc import rgb_tools
```
Does it mean that it would return different results depending if `hyperspy` is installed or not?

francisco-dlp replied:
Yes. Probably this is a bad idea. I did it to avoid passing somehow the information that it is RGB data, but on a second though, that's maybe a better approach.

Also, I've just realized that phenom and tiff still depend on hyperspy's rgb_tools. I will uncheck this task from the progress tracker above.

ericpre replied:
> Also, I've just realized that phenom and tiff still depend on hyperspy's rgb_tools. I will uncheck this task from the progress tracker above.

I notice this too but I came to the conclusion that if the data follows hyperspy structured numpy array for RGB(A), which is quite specific, it most likely come from hyperspy! As this can't be sure this is correct, this may need to change at some point and this can be done later.

It seems that this is used in various place in hyperspy, shall we could move it to the the public API of rosettasciio and used it in hyperspy?

francisco-dlp replied:
Sounds good. HyperSpy stores RGB data in struct arrays to enable seamless multi-dimensional RGB datasets. Without moving `rgb_tools` to RosettaSciIO it will be difficult for other packages to read/save RGB data in hspy/zspy format.

francisco-dlp committed:
EMD no longer requires HyperSpy for reading files

francisco-dlp committed:
Pass dictionary instead of Signal to file_writer

francisco-dlp committed:
Make MSA file writer work with dictionary

francisco-dlp commented:
With the last few commits, all reading is done without HyperSpy.

Other than making `_hierarchical` HyperSpy-free, in terms of coding only passing a dictionary to `file_writer` is left. I have started working on this in a separate branch https://github.com/francisco-dlp/hyperspy/tree/rsciio_hs_free_write. This is because changing the IO code to pass a dictionary breaks all the saving features and therefore many tests fail. I have added a progress tracker above. If somebody wants to help with this:

1. pick one of the remaining tasks from the "Pass dictionary instead of signal to file_writers" section above
2. announce your intentions here
3. Branch-out from https://github.com/francisco-dlp/hyperspy/tree/rsciio_hs_free_write and commit your changes
4. Send your PR to https://github.com/francisco-dlp/hyperspy/tree/rsciio_hs_free_write

I have already done the MSA writer.

ericpre committed:
Add test for de5 io plugin

ericpre commented:
Opened https://github.com/francisco-dlp/hyperspy/pull/49 to sort out #2803.

ericpre committed:
Improve coverage and add test save, load cycle

ericpre commented:
See https://github.com/francisco-dlp/hyperspy/pull/50 to fix breaking API of `file_writer`.

ericpre committed:
Make USID_HDF5 file writer work with dictionary

ericpre committed:
Move date_time_tools to rsciio

ericpre committed:
Make tiff writer work with dictionary

ericpre committed:
Make nexus writer work with dictionary

ericpre committed:
Make ripple writer work with dictionary

ericpre committed:
Make mrcz writer work with dictionary

ericpre committed:
Make image writer work with dictionary

francisco-dlp committed:
Merge pull request #49 from ericpre/de5_file

[SPLIT-IO] Add De5 file

francisco-dlp committed:
Merge pull request #51 from ericpre/rsciio_hs_free_write_usid

[SPLIT-IO] Make `file_writer` work with dictionary

ericpre commented:
Shall we rebase https://github.com/hyperspy/hyperspy/pull/2896 on this branch? If so, I am happy to do it.

francisco-dlp commented:
OK, thanks.

francisco-dlp committed:
Fix tvips (was not writing in hyperspy) and make it take dict

francisco-dlp committed:
Make semper writer it take dict

ericpre committed:
Move CEOS prz io plugins to rsciio

ericpre commented:
https://github.com/francisco-dlp/hyperspy/pull/52

francisco-dlp committed:
Merge pull request #52 from ericpre/SPLIT_IO-2-panta_rhei_io_plugin

[SPLIT-IO] Panta Rhei IO plugin

francisco-dlp committed:
Make blockfile writer take dict

francisco-dlp committed:
Merge remote-tracking branch 'francisco/SPLIT_IO-2' into rsciio_hs_free_write

francisco-dlp committed:
Make EMD writer take dict

francisco-dlp committed:
Make panta rhei writer take dict

francisco-dlp committed:
Move ensure_unicode to rsciio and copy get_obj to rsciio

ericpre committed:
Move docs to RosettaSciIO

ericpre committed:
Improve structure toctree

ericpre committed:
add .gitignore

ericpre commented:
I started to move the documentation.

francisco-dlp committed:
_hierarchical does not require hyperspy for reading

francisco-dlp commented:
On my side, I started working on the hspy/zspy plugins.

francisco-dlp committed:
Make hspy/zspy independent of HyperSpy

francisco-dlp committed:
Fix storing AxesManager in dict

francisco-dlp committed:
Update skimage version requirement

francisco-dlp commented:
I've just merged  https://github.com/francisco-dlp/hyperspy/tree/rsciio_hs_free_write here. All the plugins are now HyperSpy-free.

francisco-dlp committed:
Remove unnecessary import

ericpre commented:
What I have done on the docs is in https://github.com/francisco-dlp/hyperspy/pull/53. I open the PR so that it can be merged at it is and work on it later in order to get things moving. Alternatively I can work more on it later. For the docs, we are missing the RosettaScIO links (repo, docs), which means that even if we get more documentation done, we can't completely finish it and will need revisiting later.
I am fine with both approaches!

francisco-dlp committed:
Merge pull request #53 from ericpre/SPLIT_IO-docs

[SPLIO-IO] docs

francisco-dlp commented:
Thanks @ericpre. I've just merged https://github.com/francisco-dlp/hyperspy/pull/53.

The only remaining task is removing the traits dependency, and then we should be good to review and move RosettaSciIO to its own repository.

jlaehne commented:
> What I have done on the docs is in [francisco-dlp#53](https://github.com/francisco-dlp/hyperspy/pull/53). I open the PR so that it can be merged at it is and work on it later in order to get things moving. Alternatively I can work more on it later. For the docs, we are missing the RosettaScIO links (repo, docs), which means that even if we get more documentation done, we can't completely finish it and will need revisiting later. I am fine with both approaches!

Next week, I should also find some time to help with the documentation and reviewing.

ericpre commented on hyperspy/misc/io/tools.py:
```diff
@@ -1,151 +0,0 @@
-
-def overwrite(fname):
-    """ If file exists 'fname', ask for overwriting and return True or False,
-    else return True.
-
-    Parameters
-    ----------
-    fname : str or pathlib.Path
-        File to check for overwriting.
-
-    Returns
-    -------
-    bool : 
-        Whether to overwrite file.
-
-    """
-    if Path(fname).is_file() or (Path(fname).is_dir() and
-                                 os.path.splitext(fname)[1] == '.zspy'):
-        message = f"Overwrite '{fname}' (y/n)?\n"
-        try:
-            answer = input(message)
-            answer = answer.lower()
-            while (answer != "y") and (answer != "n"):
-                print("Please answer y or n.")
-                answer = input(message)
-            if answer.lower() == "y":
-                return True
-            elif answer.lower() == "n":
-                return False
-        except:
-            # We are running in the IPython notebook that does not
-            # support raw_input
-            _logger.info(
-                "Your terminal does not support raw input. "
-                "Not overwriting. "
-                "To overwrite the file use `overwrite=True`"
-            )
-            return False
-    else:
-        return True
```
Does it make sense to move these two to RosettaSciIO?

ericpre commented on hyperspy/misc/utils.py:
```diff
@@ -526,6 +527,26 @@ def as_dictionary(self):
+                            container[i] = {"_sig_" :container[i]._to_dictionary()}
+                        if to_tuple:
+                            container = tuple(container)
+                    item = container
```
Recover coverage.

ericpre commented on hyperspy/signal.py:
```diff
@@ -91,6 +91,58 @@
 except ImportError:
     CUPY_INSTALLED = False
 
+def _dic_get_hs_obj_paths(dic, axes_managers, signals, containers):
```
Add tests

ericpre commented on rosettascio/.gitignore:
```diff
@@ -0,0 +1 @@
+/docs/_build/*
```
I don't know if it matters for moving the code, but there is a typo in the folder name, it should be rosettasciio with 2 ii!

ericpre committed:
Remove traits dependence

ericpre committed:
Remove traits dependence, 2nd pass

ericpre commented:
Remove traits dependency in RosettaScIO: https://github.com/francisco-dlp/hyperspy/pull/54

jlaehne committed:
fix doc references

jlaehne committed:
introductory paragraph

jlaehne committed:
fixes

jlaehne committed:
supported formats intro

jlaehne committed:
references in CHANGES.rst

jlaehne committed:
sort supported formats, add extension

jlaehne committed:
link supported formats

francisco-dlp committed:
Merge pull request #54 from ericpre/SPLIT_IO-remove_traits_dependence

[SPLIT-IO] remove traits dependence

francisco-dlp commented:
Merged, thanks @ericpre!

@jlaehne  is currently working on the documentation in https://github.com/francisco-dlp/hyperspy/pull/55 . Once that's ready, and the code reviewed, I think that we can merge this into RnM and create a repository for RosettaSciIO.

What would be the most effective way of letting contributors know that any IO development should happen in RosettaSciIO?

jlaehne committed:
consistency, format titles

jlaehne committed:
consistency semper

jlaehne committed:
replace HyperSpy mentions where appropriate

jlaehne committed:
update interoperability

jlaehne committed:
fixes

jlaehne commented:
> @jlaehne is currently working on the documentation in [francisco-dlp#55](https://github.com/francisco-dlp/hyperspy/pull/55) . Once that's ready, and the code reviewed, I think that we can merge this into RnM and create a repository for RosettaSciIO.

Working on the documentation should not keep us from reviewing the code ;-)
Anyway, the documentation changes may be merged at any point and then continued in the new repository.

> What would be the most effective way of letting contributors know that any IO development should happen in RosettaSciIO?

I will see where to add something to the HyperSpy documentation. Otherwise, I guess we should announce it on gitter. Maybe we could make a first post there already, inviting people to review the code.

jlaehne committed:
typos interoperability

jlaehne committed:
update hyperspy docs

jlaehne committed:
update issue tracker links

jlaehne committed:
fix

jlaehne committed:
add note to dev-guide

jlaehne committed:
pass not needed

jlaehne committed:
unneccesary imports

jlaehne committed:
unneccesary assignment

jlaehne committed:
fix tiff

jlaehne committed:
Merge branch 'SPLIT_IO-2' into rsciio-lgtm

ericpre reviewed (commented):
I think that this is pretty much ready. Maybe the only thing that would be good doing now is adding the test to `_dic_get_hs_obj_paths` and `_obj_in_dict2hspy` in hyperspy to check that it is working as expected. Other comments can be addressed after the split.

francisco-dlp committed:
Merge pull request #55 from jlaehne/rosetta-doc-test

Fix references in documentation, ...

francisco-dlp committed:
Merge pull request #56 from jlaehne/rsciio-lgtm

rsciio lgtm consolidation

jlaehne commented:
I have been running a number of old notebooks reading different file formats on my local machine as a cross-check and it works ;-)

jlaehne committed:
mention RosettaSciIO in getting_started

jlaehne commented:
Just added another minor doc improvement in https://github.com/francisco-dlp/hyperspy/pull/57

francisco-dlp commented:
Thanks @ericpre for the review. I am busy today and tomorrow and I won't be able to finish this before leaving on holidays on Saturday. @ericpre and @jlaehne, if any of you wants to finish finish this, merge it and split the repository, go for it. Otherwise, I can take care of it in late August.

Remaining tasks (from @ericpre's review):
* [x] Correct typo in folder name
* [x] Fully remove `rbg_tools`
* [x] Add tests for:
    * [x] `_obj_in_dict2hspy`
    * [x] `_dic_get_hs_obj_paths`
* [x] Merge
* [x] Split

In my experience, following [the official github instructions](https://docs.github.com/en/get-started/using-git/splitting-a-subfolder-out-into-a-new-repository) for splitting a repository works well.

francisco-dlp committed:
Merge pull request #57 from jlaehne/rsciio-link

mention RosettaSciIO in getting_started

ericpre commented:
Considering that there is not much left to do, I would prefer to finish in the next few days. After that, I am happy to give a go at splitting! :)

jlaehne commented:
> Considering that there is not much left to do, I would prefer to finish in the next few days. After that, I am happy to give a go at splitting! :)

+1 - would be great if you could do so!

@pietsjoh will be starting to contribute new readers in the next weeks and it would be great to do so directly to the new repo.

P.S.: I am available until Tuesday to review the remaining steps.

ericpre committed:
Remove folder from rosettascio to rosettasciio

ericpre committed:
Move rgbtools to rosettasciio

ericpre committed:
Fix package name "RosettaSciIO"

ericpre committed:
Remove traits dependency

ericpre committed:
Add test for `_obj_in_dict2hspy`

ericpre commented:
The remaining tasks are done in https://github.com/francisco-dlp/hyperspy/pull/58.

@francisco-dlp, just in case, we can merge and push to your branch, would have time to review and merge https://github.com/francisco-dlp/hyperspy/pull/58? The change are mostly trivial.

Worst-case scenario, we create another branch but it would be better to keep it all in this PR!

jlaehne committed:
Update tests.yml

jlaehne committed:
Update azure-pipelines.yml

ericpre committed:
Merge pull request #22 from jlaehne/patch-1

Update tests.yml

francisco-dlp committed:
Merge pull request #58 from ericpre/SPLIT_IO-2_final

[SPLIT-IO] last few tasks

jlaehne commented:
Just realized that

1.  We are missing a changelog entry for this PR.
2.  v1.7.1 was not automatically merged from RnP into RnM and RnM is still on 1.7.1.dev0 - including the changelog.

I would propose merging this one nevertheless and sorting out anything related to the v1.7.1 merge into RnM in a separate PR, possibly including the changelog entry of this PR (which should actually be about the split and not about the preparation of the split).

ericpre commented:
Yes and there is some tidying up needed on the RELEASE_next_major branch too!

ericpre commented:
For the record, here are the step that I did to split into a separate repository:

## Command to split the repository
```
`git filter-repo --path rosettascio/ --path rosettasciio/ --path hyperspy/io_plugins/ --path hyperspy/misc/io/ --path hyperspy/tests/io/ --path doc/user_guide/io.rst --path-rename rosettasciio/: --force
```
I needed to add all relevant paths to keep the history after folder renames, etc.

## Tidy up left over code
Since it was necessary to also filter `doc/user_guide/io.rst` to keep the history, I remove it in a commit.

## Clean some large file from the history
We had some large which were added and removed from code during a pull request and never cleaned up. This was possible to find them using the following command:
```
git rev-list --objects --all |
  git cat-file --batch-check='%(objecttype) %(objectname) %(objectsize) %(rest)' |
  sed -n 's/^blob //p' |
  sort --numeric-sort --key=2 |
  cut -c 1-12,41- |
  $(command -v gnumfmt || echo numfmt) --field=2 --to=iec-i --suffix=B --padding=7 --round=nearest
```
ref: https://stackoverflow.com/questions/10622179/how-to-find-identify-large-commits-in-git-history

The largest file before history cleaning were:
```
3c272c431de2  8.7MiB rsciio/tests/emd_files/fei_emd_files.zip
b4dce2fe8c67   10MiB hyperspy/tests/io/sur_data/test_spectral_map_compressed.sur
5fdd25abfa0b   11MiB hyperspy/tests/io/edax_files/spd_map.spd.gz
f89958aa88ac   11MiB hyperspy/tests/io/edax_files.zip
48e1eba14394   21MiB hyperspy/tests/io/nexus_files/file2.nxs
b949accc2fcf   32MiB hyperspy/tests/io/sur_data/test_spectral_map.sur
4a8123661a42   35MiB hyperspy/tests/io/edax_files.zip
502c66aed3c6   35MiB rsciio/tests/edax_files.zip
e5ad43b77116   56MiB hyperspy/tests/io/nexus_files/file1.nxs
```
And are now
```
3c272c431de2  8.7MiB rsciio/tests/emd_files/fei_emd_files.zip
5fdd25abfa0b   11MiB hyperspy/tests/io/edax_files/spd_map.spd.gz
f89958aa88ac   11MiB hyperspy/tests/io/edax_files.zip
4a8123661a42   35MiB hyperspy/tests/io/edax_files.zip
502c66aed3c6   35MiB rsciio/tests/edax_files.zip
```

I keep a local copy of the git repository before in case something was incorrect! After hyperspy 2.0 is released, we should possibly do the same to the hyperspy repository.

jlaehne commented:
Thanks Eric for finishing it up!
